Correctly handle rendering/labelling polygon features from vector tiles using a point symbol

nyalldawson opened the pull request:
This is permitted in JS vector tile libraries, and is treated as the centroid of the polygon

Fixes #46587

nyalldawson committed:
Correctly handle rendering/labelling polygon features from vector tiles using
a point symbol

This is permitted in JS vector tile libraries, and is treated as
the centroid of the polygon

Fixes #46587

pathmapper commented:
Using the centroid, the icon/label is not always guaranteed to be on the polygon surface (e.g. for some concave polygons), right?

JS vector tile libraries (_Mapbox GL JS_, _Maplibre GL JS_) are using 

> an approximation of a polygon's Pole Of Inaccessibility https://en.wikipedia.org/wiki/Pole_of_inaccessibility

https://github.com/mapbox/mapbox-gl-js/blob/ade9271dc97df2f21c48be07e940b50fd0119c38/src/symbol/symbol_layout.js#L14
https://github.com/maplibre/maplibre-gl-js/blob/be9d6cb6f6935f3383d746990d3245d30d7924e8/src/symbol/symbol_layout.ts#L13

https://github.com/mapbox/mapbox-gl-js/blob/main/src/util/find_pole_of_inaccessibility.js
https://github.com/maplibre/maplibre-gl-js/blob/main/src/util/find_pole_of_inaccessibility.ts

> This is a copy of http://github.com/mapbox/polylabel adapted to use Points

There's already a polylabel implementation in QGIS available which maybe could be used here instead of the centroid:
https://github.com/qgis/QGIS/blob/fdc895e1e361463025aa37dd2b876dab5ab79869/src/core/geometry/qgsinternalgeometryengine.cpp#L244-L246

Processing tool:
https://github.com/qgis/QGIS/blob/32c2cea54cb92bbb2243b222816c8154c2b9adf9/src/analysis/processing/qgsalgorithmpoleofinaccessibility.cpp

nyalldawson committed:
When creating simple vector tile rules, default to using a
  geometry_type($geometry)='Polygon' (or line, or marker)
filter rule

The vector tile specs don't mandate that a marker or line
is only used to render a point/line feature, and instead
convey that a marker should be rendered at the center
of a polygon and a line should be used for the outline
of a polygon.

So in order to restrict markers/lines to ONLY being rendered
for point/line features, we need to use an explicit filter

nyalldawson commented:
Ok, this has now been reworked so that we correctly follow the vector tile style specs without breaking older workflows
